### PR TITLE
✨ [Tasks] JSON Schema spec for Inference types + TS type generation

### PR DESCRIPTION
**Pull request by SBrandeis**
# TL;DR

- Introduces specification for inference input/output types for each task, in JSON Schema format
- Add a "script" to generate typescript code from that JSON spec using [`quicktype-core`](https://github.com/glideapps/quicktype/tree/master/packages/quicktype-core)
    - Introduces `quicktype-core` as a dev dependency  (from our fork of quicktype https://github.com/huggingface/quicktype/releases/tag/pack-18.0.15)
    - License: Apache-2.0

# TODO

- [x] Add a `text2text-generation` task to serve as a "canonical reference" for `summarization` & `translation`
- [x] Add a `text-to-audio` task to server as a "canonical reference" for `text-to-speech`
- [ ] Code generation tweaks
   - [x] Change `any` types to `unknown`
   - [x] Generate output types for `sentence-similarity`
   - [ ] Generate output types for `feature-extraction` -> Let's do that later?
- [x] add placeholder specs

### CODE REVIEW

**Comment by SBrandeis** [comment]
Ping @coyotte508 for visibility

**Review by julien-c — APPROVED** [review]
looks 😍😍😍

File: packages/tasks/src/tasks/audio-classification/inference.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/**
+ * Inference code generated from the JSON schema spec in ./spec
+ *
+ * Generated on 2024-01-19T14:59:10.562Z
+ */
+
+/**
+ * Inputs for Audio Classification inference
+ */
+export interface AudioClassificationInput {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/audio-classification/inference.ts:10]: Audio data is usually pass through `data`

https://huggingface.co/docs/api-inference/detailed_parameters#audio-classification-task

**Review by osanseviero — COMMENTED** [review]
Nice!

File: packages/tasks/src/tasks/index.ts [hunk 2]
```diff
@@ -216,6 +216,7 @@ export interface TaskData {
 	datasets: ExampleRepo[];
 	demo: TaskDemo;
 	id: PipelineType;
+	canonicalId?: PipelineType;
```
Comment (SBrandeis) [inline packages/tasks/src/tasks/index.ts:219]: Added this property to express one task being a "subtask" of another (eg, summarization being a subtask of text2text-generation)

File: packages/tasks/src/tasks/feature-extraction/inference.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ */
+
+/**
+ * Inputs for Feature Extraction inference
+ */
+export interface FeatureExtractionInput {
+	/**
+	 * One or several texts to get the features of
+	 */
+	inputs: string[] | string;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/feature-extraction/inference.ts:14]: We need to decide if we want to support batched inputs or not. Community API does not support batched inputs, Inference API does support them.

File: packages/tasks/src/tasks/feature-extraction/inference.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Inference code generated from the JSON schema spec in ./spec
+ *
+ * Using src/scripts/inference-codegen
+ */
+
+/**
+ * Inputs for Feature Extraction inference
+ */
+export interface FeatureExtractionInput {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/feature-extraction/inference.ts:10]: Btw not sure how to communicate about it, but we should consider at one point renaming to "Embedding" as that's the name used by the community nowadays. This endpoint retrieves sentence embeddings if it's a ST repo, or contextualized word embeddings if it's a TF repo

File: packages/tasks/src/tasks/feature-extraction/spec/output.json [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{
+	"$id": "/inference/schemas/feature-extraction/output.json",
+	"$schema": "http://json-schema.org/draft-06/schema#",
```
Comment (osanseviero) [inline packages/tasks/src/tasks/feature-extraction/spec/output.json:3]: This one was not exported I think

File: packages/tasks/src/tasks/image-classification/inference.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Image Classification
+ */
+export interface ImageClassificationParameters {
+	/**
+	 * When specified, limits the output to the top K most probable classes.
+	 */
+	topK?: number;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-classification/inference.ts:31]: I would also add `function_to_apply` so people can apply `sigmoid`/`softmax`.

File: packages/tasks/src/tasks/image-classification/inference.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+	[property: string]: unknown;
+}
+
+/**
+ * Outputs of inference for the Image Classification task
+ */
+export interface ImageClassificationOutput {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-classification/inference.ts:38]: I think the output is a list of these

File: packages/tasks/src/tasks/image-segmentation/inference.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+ */
+export interface ImageSegmentationOutput {
+	/**
+	 * The label of the predicted segment
+	 */
+	label: string;
+	/**
+	 * The corresponding mask as a black-and-white image
+	 */
+	mask: unknown;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-segmentation/inference.ts:62]: Also the score (optionally) https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/pipelines/image_segmentation.py#L136

File: packages/tasks/src/tasks/image-segmentation/inference.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+export type ImageSegmentationSubtask = "instance" | "panoptic" | "semantic";
+
+/**
+ * Outputs of inference for the Image Segmentation task
+ *
+ * A predicted mask / segment
+ */
+export interface ImageSegmentationOutput {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-segmentation/inference.ts:54]: It's a list of these dicts (if we go with one sample) or list of list of dicts (if we do batching)

File: packages/tasks/src/tasks/image-to-image/inference.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+export type ImageToImageOutput = unknown[];
+
+/**
+ * Inputs for Image To Image inference
+ */
+export interface ImageToImageInput {
+	/**
+	 * One or more images to generate images from
+	 */
+	inputs: unknown;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-to-image/inference.ts:16]: just as audio, my understanding is that the image is passed through data rather than params

File: packages/tasks/src/tasks/image-to-image/inference.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+	/**
+	 * Additional inference parameters
+	 */
+	parameters?: unknown;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/image-to-image/inference.ts:20]: FYI there are some params only from diffusers here https://github.com/huggingface/api-inference-community/blob/main/docker_images/diffusers/app/pipelines/image_to_image.py

* num_inference_steps
* negative_prompt
* guidance_scale

**Review by osanseviero — COMMENTED** [review]
Very cool! I did a quick high-level pass

File: packages/tasks/src/tasks/audio-classification/inference.ts (hunk 1 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/audio-classification/inference.ts:10]: Same for images. The jsonschema cannot specify this since sending as raw data and sending as json are 2 different things. So for now it's kind of a blind spot. If we provide an openapi schema for our APIs in the future, then it will be possible to document it. Openapi easily integrates with jsonschema so having them is already a first good step.

(difference between a jsonschema as in this PR and an openapi description is that this PR describes objects with their attributes while the openapi description with include stuff like server routes, accepted headers, etc.)

File: packages/tasks/src/tasks/audio-classification/inference.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (Wauplin) [inline packages/tasks/src/tasks/audio-classification/inference.ts:1]: In term of repo structure, I would put all the `.ts` generated files inside a `./generated` folder to make it explicit to any reader that the files are generated.

(typically comments from @osanseviero should be made on the `input.json`/`output.json` files instead of `inference.ts`).

File: packages/tasks/src/tasks/placeholder/spec/input.json [hunk 13]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (Wauplin) [inline packages/tasks/src/tasks/placeholder/spec/input.json:1]: Is `placeholder` intended to be committed?

**Review by Wauplin — COMMENTED** [review]
:fire: :fire:

I've commented with just some high-level feedback. Do you want a review of each task individually as well or is this PR mostly to approve the concept? (asking before diving in a longer review)

File: packages/tasks/src/tasks/feature-extraction/spec/output.json (hunk 5 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/feature-extraction/spec/output.json:3]: Yes - it's an issue with the package we use for code generation, I don't know how to fix it yet

File: packages/tasks/src/tasks/document-question-answering/inference.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+export interface DocumentQuestionAnsweringInpu {
+	/**
+	 * The image on which the question is asked
+	 */
+	image?: unknown;
+	/**
+	 * A question to ask of the document
+	 */
+	question?: string;
```
[the remark this answers is not in the document]
Comment (SBrandeis) [reply packages/tasks/src/tasks/document-question-answering/inference.ts]: good catch

File: packages/tasks/src/tasks/document-question-answering/inference.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	answer: string;
+	end: number;
+	/**
+	 * The probability associated to the answer.
+	 */
+	score: number;
+	start: number;
+	/**
+	 * The index of each word/box pair that is in the answer
+	 */
```
[the remark this answers is not in the document]
Comment (SBrandeis) [reply packages/tasks/src/tasks/document-question-answering/inference.ts:98]: The fields are ordered in alphabetical order by the generation script, is that what you are suggesting to change here?

File: packages/tasks/src/tasks/image-classification/inference.ts (hunk 7 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/image-classification/inference.ts:44]: Yes - it should be outputed as such by the code generation tool as the spec specifies correctly this should be an array

Same issue for `FeatureExtraction` output not being generated - I'll look into it

File: packages/tasks/src/tasks/audio-classification/inference.ts (hunk 1 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/audio-classification/inference.ts:10]: Yes - sorry for the delay in answering

Leaving the image/audio data as `unknown` was intentional, to give more flexibility to the libraries.
Image & audio data can be passed in several different forms (raw binary data, path to a local or remote file, base64 encoded data...) and I did not want to constrain downstream users of those types into one single representation.

> (difference between a jsonschema as in this PR and an openapi description is that this PR describes objects with their attributes while the openapi description with include stuff like server routes, accepted headers, etc.)

Yes that is correct, there will be some additional work necessary to generate an OpenAPI spec for an inference API (including actually specifying how we expect the binary data to be represented)

**Comment by SBrandeis** [comment]
@Wauplin I think we have enough validation of the concept to start a more in-depth review of the specs and generated code 👍

File: packages/tasks/src/tasks/placeholder/spec/input.json (hunk 13 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/placeholder/spec/input.json:1]: Yes, in case we ever have to add a new task to this package

**Comment by Wauplin** [comment]
> I think we have enough validation of the concept to start a more in-depth review of the specs and generated code 👍

Great, I'll review that on tomorrow then!

File: packages/tasks/src/tasks/audio-classification/inference.ts (hunk 12 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/audio-classification/inference.ts:1]: Hmm, I kindof like the fact that you can import types from `tasks/text-generation/inference` directly, and that the inference types are close to the JSON Schema spec

File: packages/tasks/src/tasks/image-to-image/inference.ts (hunk 11 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/image-to-image/inference.ts:20]: Thanks for the pointer!

File: packages/tasks/src/tasks/audio-classification/spec/input.json [hunk 16]
```diff
@@ -0,0 +1,29 @@
+{
+	"$id": "/inference/schemas/audio-classification/input.json",
+	"$schema": "http://json-schema.org/draft-06/schema#",
```
Comment (Wauplin) [inline packages/tasks/src/tasks/audio-classification/spec/input.json:3]:
```suggestion
	"$schema": "http://json-schema.org/draft-06/schema#",
        "title": "AudioClassificationInput",
```

(nit) missing `"title"` field in the input specs. How does quicktype know which name to give to the dataclass? (for output specs it seems `"title"` is there)

File: packages/tasks/src/tasks/depth-estimation/spec/input.json [hunk 17]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (Wauplin) [inline packages/tasks/src/tasks/depth-estimation/spec/input.json:1]: Note: depth estimation is not listed in https://huggingface.co/docs/api-inference/detailed_parameters + widget doesn't exist. Should we still list it in case we add it in the future?

File: packages/tasks/src/tasks/depth-estimation/spec/output.json [hunk 18]
```diff
@@ -0,0 +1,9 @@
+{
+	"$id": "/inference/schemas/depth-estimation/output.json",
+	"$schema": "http://json-schema.org/draft-06/schema#",
+	"description": "Outputs of inference for the Depth Estimation task",
+	"type": "array",
+	"items": {
+		"description": "The output depth labels"
+	}
```
Comment (Wauplin) [inline packages/tasks/src/tasks/depth-estimation/spec/output.json:8]: we don't know the output format I guess?

File: packages/tasks/src/tasks/document-question-answering/inference.ts [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+ */
+
+/**
+ * Inputs for Document Question Answering inference
+ */
+export interface DocumentQuestionAnsweringInput {
+	/**
+	 * One or several document+question pairs to answer
+	 */
+	inputs: DocumentQuestionAnsweringInpu[] | DocumentQuestionAnsweringInpu;
```
Comment (Wauplin) [inline packages/tasks/src/tasks/document-question-answering/inference.ts:14]: `DocumentQuestionAnsweringInpu` truncating the name to declare the second class is done by quicktype I guess. Would it be possible to rename it to `DocumentQuestionAnsweringInputItem` for example? (by adding an explicit `title` field)

File: packages/tasks/src/tasks/document-question-answering/spec/input.json [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+					}
+				}
+			}
+		},
+		"DocumentAndQuestion": {
+			"type": "object",
+			"properties": {
+				"image": {
+					"description": "The image on which the question is asked"
+				},
```
Comment (Wauplin) [inline packages/tasks/src/tasks/document-question-answering/spec/input.json:86]:
```suggestion
				"image": {
					"description": "The image on which the question is asked, encoded as a base64 string."
					"type": "string",
				},
```

maybe mention that image must be a base64-encoded string?

File: packages/tasks/src/tasks/feature-extraction/inference.ts [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+	/**
+	 * One or several texts to get the features of
+	 */
+	inputs: string[] | string;
+	/**
+	 * Additional inference parameters
+	 */
+	parameters?: { [key: string]: unknown };
+	[property: string]: unknown;
```
Comment (Wauplin) [inline packages/tasks/src/tasks/feature-extraction/inference.ts:19]: Text Embedding Inference supports both `truncate` and `normalize` boolean parameters (see [here](https://huggingface.github.io/text-embeddings-inference/#/Text%20Embeddings%20Inference/embed)). It has been asked twice in `huggingface_hub` but for now I'm waiting for the "official" spec before adding them. Should we add them now (I'm in favor of it since TEI should become the "flagship" for embeddings)

File: packages/tasks/src/tasks/text-generation/inference.ts [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+	parameters?: TextGenerationParameters;
+	[property: string]: unknown;
+}
+
+/**
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Text Generation
+ */
+export interface TextGenerationParameters {
```
Comment (Wauplin) [inline packages/tasks/src/tasks/text-generation/inference.ts:27]: TGI also have a `details: bool` parameters to return more details from the process. It also affects the output type.

At the moment we have quite a hacky way to handle both TGI and non-TGI servers in `huggingface_hub.InferenceClient`. What I do is to send TGI-specific parameters to the server and if I receive an error, I retry without the TGI-specific parameters. See [here](https://github.com/huggingface/huggingface_hub/blob/7bb8308909d0981ec911d48a52c16b7ba2c595fc/src/huggingface_hub/inference/_client.py#L1484) on how it's done.

Having a unified API specification should help here. I think it makes sense to have as much parameters as possible in the spec (or at least have the ones for TGI) and if the server does not support them, it ignores them (maybe send back a warning in the output?)

**Review by Wauplin — COMMENTED** [review]
I started by reviewing some of the tasks I know more (because of trickiness to handle in huggingface_hub). I'm starting with that and will continue on the other tasks tomorrow :)

File: packages/tasks/src/tasks/audio-classification/spec/input.json (hunk 16 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/audio-classification/spec/input.json:3]: > How does quicktype know which name to give to the dataclass? 

I think it does some inference based on the `name` attribute we set here: https://github.com/huggingface/huggingface.js/blob/a33987fe219616a75c0e9fca79a65164472a9dd0/packages/tasks/src/scripts/inference-codegen.ts#L46

File: packages/tasks/src/tasks/document-question-answering/inference.ts (hunk 19 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/document-question-answering/inference.ts]: Handled in 6558af4 - good catch!

Comment (SBrandeis) [reply packages/tasks/src/tasks/document-question-answering/inference.ts]: According to this doc: https://huggingface.co/docs/transformers/v4.37.1/en/main_classes/pipelines#transformers.DepthEstimationPipeline

The output is an image, so I specced it with the same type as image inputs

File: packages/tasks/src/tasks/depth-estimation/spec/output.json (hunk 18 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/depth-estimation/spec/output.json:8]: According to this doc: https://huggingface.co/docs/transformers/v4.37.1/en/main_classes/pipelines#transformers.DepthEstimationPipeline

The output is an image, so I specced it with the same type as image inputs

File: packages/tasks/src/tasks/document-question-answering/spec/input.json (hunk 20 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/document-question-answering/spec/input.json:87]: I'm not sure we want to enforce the image format at that level in the spec

IMO it would be better to leave image & audio data as "untyped" whenever possible to allow dependents to use whatever they see fit as an image type

File: packages/tasks/src/tasks/feature-extraction/inference.ts (hunk 3 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/feature-extraction/inference.ts:14]: Maybe let's keep it simple for those types and remove batching then?

File: packages/tasks/src/tasks/feature-extraction/inference.ts (hunk 4 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/feature-extraction/inference.ts:10]: How would we proceed to do this?

File: packages/tasks/package.json [hunk 23]
```diff
@@ -24,9 +24,10 @@
 		"format": "prettier --write .",
 		"format:check": "prettier --check .",
 		"prepublishOnly": "pnpm run build",
-		"build": "tsup src/index.ts --format cjs,esm --clean --dts",
+		"build": "tsup src/index.ts src/scripts/**.ts --format cjs,esm --clean --dts",
```
[the remark this answers is not in the document]
Comment (SBrandeis) [reply packages/tasks/package.json:27]: ^ that command fails with a syntax error (`cannot use import outside of a module` or smth)

Is it OK tu use `tsc` to compile the script in `pnpm run inference-codegen`?

File: packages/tasks/src/scripts/inference-codegen.ts [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+		try {
+			require(`${currentPath}/package.json`);
+			return path.normalize(currentPath);
+		} catch (err) {
+			/// noop
+		}
+		level--;
+	}
+	return "";
+};
```
[the remark this answers is not in the document]
Comment (SBrandeis) [reply packages/tasks/src/scripts/inference-codegen.ts:31]: It's incompatible with `tsc` it seems?

```
~/Documents/dev/hf/huggingface.js/packages/tasks (tasks-specifications) > tsc
scripts/inference-codegen.ts:9:20 - error TS1343: The 'import.meta' meta-property is only allowed when the '--module' option is 'es2020', 'es2022', 'esnext', 'system', 'node16', or 'nodenext'.
```

File: packages/tasks/src/tasks/audio-classification/spec/input.json (hunk 16 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/audio-classification/spec/input.json:3]: If possible I would explicitly add titles everywhere so that we stay consistent in case we get rid of quicktype at some point or if quicktype implicit naming rules change.

Comment (Wauplin) [reply packages/tasks/src/tasks/audio-classification/spec/input.json:3]: (in case it can help, I used this dirty script to update all of the schemas [in this PR](https://github.com/huggingface/inference-schemas/pull/4/files#diff-c02135df8f023a5f1631042fe89eb6454763fbeaa84698fa250a6912813be2db):

```py
from pathlib import Path

schemas_dir = Path("./schemas")

for schema_file in schemas_dir.glob("**/*.json"):
    task = schema_file.parent.name
    if "input" in str(schema_file):
        title = "".join(part.capitalize() for part in task.split("-")) + "Input"
    else:
        title = "".join(part.capitalize() for part in task.split("-")) + "Output"

    with open(schema_file, "r") as f:
        schema_lines = f.readlines()

    schema_lines.insert(4, f'    "title": "{title}",\n')
    with open(schema_file, "w") as f:
        f.writelines(schema_lines)
```
)

File: packages/tasks/src/tasks/document-question-answering/spec/input.json (hunk 20 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/document-question-answering/spec/input.json:87]: Right, makes sense :+1:

File: packages/tasks/src/tasks/depth-estimation/spec/output.json (hunk 18 above)
Comment (Wauplin) [reply packages/tasks/src/tasks/depth-estimation/spec/output.json:8]: Right, didn't know that :+1: So maybe update the description to something like `Output image representing the estimated depth.`?

It seems that the `transformers`'s dostring for [DepthEstimationPipeline](https://github.com/huggingface/transformers/blob/7fa4b36ebaebc6f5ca72ccf08ed0fe2a84422c5c/src/transformers/pipelines/depth_estimation.py#L76) is an exact copy-paste from [ImageClassificationPipeline](https://github.com/huggingface/transformers/blob/7fa4b36ebaebc6f5ca72ccf08ed0fe2a84422c5c/src/transformers/pipelines/image_classification.py#L149) (depth estimation shouldn't be returning labels _I guess_). (cc @amyeroberts maybe?)

**Comment by SBrandeis** [comment]
I opened an issue in the upstream repo regarding top-level array generation
https://github.com/glideapps/quicktype/issues/2481

File: packages/tasks/src/tasks/audio-classification/spec/input.json (hunk 16 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/audio-classification/spec/input.json:3]: Done - thanks!

**Comment by SBrandeis** [comment]
I added a "post-process" script using the typescript API to generate the appropriate array type while https://github.com/glideapps/quicktype/issues/2481 is being handled

File: packages/tasks/src/tasks/audio-classification/inference.ts [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Audio Classification
+ */
+export interface AudioClassificationParameters {
+	/**
+	 * The function to apply to the model outputs in order to retrieve the scores.
+	 */
+	functionToApply?: AudioClassificationOutputTransform;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/audio-classification/inference.ts:29]: this is not supported by any library afaik

File: packages/tasks/src/tasks/automatic-speech-recognition/inference.ts [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+ */
+export interface AutomaticSpeechRecognitionInput {
+	/**
+	 * The input audio data
+	 */
+	data: unknown;
+	/**
+	 * Additional inference parameters
+	 */
+	parameters?: {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/automatic-speech-recognition/inference.ts:17]: `return_timestamps` is supported by a bunch of models.
For Whisper, you have many of the generative params as well

Comment (osanseviero) [reply packages/tasks/src/tasks/automatic-speech-recognition/inference.ts:17]: See https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.AutomaticSpeechRecognitionPipeline

File: packages/tasks/src/tasks/automatic-speech-recognition/inference.ts [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+		[key: string]: unknown;
+	};
+	[property: string]: unknown;
+}
+export type AutomaticSpeechRecognitionOutput = AutomaticSpeechRecognitionOutputElement[];
+/**
+ * Outputs of inference for the Automatic Speech Recognition task
+ */
+export interface AutomaticSpeechRecognitionOutputElement {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/automatic-speech-recognition/inference.ts:26]: If we use `return_timestamps`, then this also contains `chunks`

File: packages/tasks/src/tasks/depth-estimation/spec/output.json (hunk 18 above)
Comment (osanseviero) [reply packages/tasks/src/tasks/depth-estimation/spec/output.json:9]: Iirc, the output is a dictionary with two entries, one being the `depth` which is a depth estimation image, the other is `predicted_depth`, which is the tensor. See https://huggingface.co/docs/transformers/main/tasks/monocular_depth_estimation

Comment (osanseviero) [reply packages/tasks/src/tasks/depth-estimation/spec/output.json:9]: https://github.com/huggingface/transformers/issues/28729

File: packages/tasks/src/tasks/document-question-answering/inference.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (osanseviero) [reply packages/tasks/src/tasks/document-question-answering/inference.ts:99]: I guess the alphabetical order is a bit weird with the docstrings. We have `The answer to the question.`, then answer, then end, much later start.

```
/**
* The answer to the question.
*/
answer: string;
end: number;
/**
* The probability associated to the answer.
*/
score: number;
start: number;
/**
```

File: packages/tasks/src/tasks/feature-extraction/inference.ts (hunk 21 above)
Comment (osanseviero) [reply packages/tasks/src/tasks/feature-extraction/inference.ts:21]: None of this would work out of the box in `sentence_transformers` API, but I guess we can add later on if needed

File: packages/tasks/src/tasks/feature-extraction/spec/output.json [hunk 28]
```diff
@@ -0,0 +1,7 @@
...
+	"description": "The embedding for the input text, as a nested list (tensor) of floats",
+	"type": "array",
```
Comment (osanseviero) [inline packages/tasks/src/tasks/feature-extraction/spec/output.json:5]: Note: it's an array in sentence transformers (one embedding per input), a list within a list in transformers (one embedding per token), and a list within a list within a list in Inference API (for batching) iirc


1. https://github.com/huggingface/api-inference-community/blob/main/docker_images/sentence_transformers/app/pipelines/feature_extraction.py#L25
2. https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/feature_extraction.py#L96

File: packages/tasks/src/tasks/sentence-similarity/spec/output.json [hunk 29]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (osanseviero) [inline packages/tasks/src/tasks/sentence-similarity/spec/output.json:1]: Note that this one is not exported

File: packages/tasks/src/tasks/table-question-answering/inference.ts [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+export interface TableQuestionAnsweringInputData {
+	/**
+	 * The question to be answered about the table
+	 */
+	question: string;
+	/**
+	 * The table to serve as context for the questions
+	 */
+	table: {
+		[key: string]: unknown;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/table-question-answering/inference.ts:34]: the values are list of strings iirc

File: packages/tasks/src/tasks/text-to-audio/inference.ts [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Text To Audio
+ */
+export interface TextToAudioParameters {
+	/**
+	 * Parametrization of the text generation process
+	 */
+	generate?: GenerationParameters;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/text-to-audio/inference.ts:29]: We should also support forward params so we can pass things such as `speaker_embeddings` in SpeechT5 https://huggingface.co/microsoft/speecht5_tts

File: packages/tasks/src/tasks/text-to-audio/inference.ts [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+	[property: string]: unknown;
+}
+/**
+ * Parametrization of the text generation process
+ *
+ * Ad-hoc parametrization of the text generation process
+ */
+export interface GenerationParameters {
+	/**
+	 * I can be the papa you'd be the mama
```
Comment (osanseviero) [inline packages/tasks/src/tasks/text-to-audio/inference.ts:39]: :smile:

File: packages/tasks/src/tasks/text-to-audio/inference.ts [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+	 */
+	temperature?: number;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/text-to-audio/inference.ts:41]: Should we add the others?

File: packages/tasks/src/tasks/text-to-image/inference.ts [hunk 34]
```diff
@@ -0,0 +1,69 @@
...
+	parameters?: TextToImageParameters;
+	[property: string]: unknown;
+}
+
+/**
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Text To Image
+ */
+export interface TextToImageParameters {
```
Comment (osanseviero) [inline packages/tasks/src/tasks/text-to-image/inference.ts:27]: Missing `scheduler` I think

File: packages/tasks/src/tasks/text2text-generation/inference.ts [hunk 35]
```diff
@@ -0,0 +1,53 @@
+/**
+ * Inference code generated from the JSON schema spec in ./spec
```
Comment (osanseviero) [inline packages/tasks/src/tasks/text2text-generation/inference.ts:2]: Should we use this opportunity to unify `text-generation` and `text2text-generation`?

File: packages/tasks/src/tasks/automatic-speech-recognition/inference.ts (hunk 26 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/automatic-speech-recognition/inference.ts]: good catch, thanks!

File: packages/tasks/src/tasks/token-classification/inference.ts [hunk 36]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Additional inference parameters
+ *
+ * Additional inference parameters for Token Classification
+ */
+export interface TokenClassificationParameters {
+	/**
+	 * The strategy used to fuse tokens based on model predictions
+	 */
+	aggregationStrategy?: TokenClassificationAggregationStrategy;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/token-classification/inference.ts:29]: A bit strange as this is actually a load parameter, not an inference parameter - see https://huggingface.co/docs/transformers/main/en/main_classes/pipelines#transformers.TextToAudioPipeline

File: packages/tasks/src/tasks/video-classification/inference.ts [hunk 37]
```diff
@@ -0,0 +1,59 @@
...
+	 * The number of sampled frames to consider for classification.
+	 */
+	numFrames?: number;
+	/**
+	 * When specified, limits the output to the top K most probable classes.
+	 */
+	topK?: number;
+	[property: string]: unknown;
+}
+export type TextClassificationOutputTransform = "sigmoid" | "softmax" | "none";
```
Comment (osanseviero) [inline packages/tasks/src/tasks/video-classification/inference.ts:44]: Maybe let's rename to a more general class (`ClassificationOutputTransform`?)

File: packages/tasks/src/tasks/visual-question-answering/inference.ts [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+export type VisualQuestionAnsweringOutput = VisualQuestionAnsweringOutputElement[];
+/**
+ * Outputs of inference for the Visual Question Answering task
+ */
+export interface VisualQuestionAnsweringOutputElement {
+	/**
+	 * The answer to the question
+	 */
+	answer?: string;
+	label: unknown;
```
Comment (osanseviero) [inline packages/tasks/src/tasks/visual-question-answering/inference.ts:57]: this is a string

File: packages/tasks/src/tasks/zero-shot-object-detection/spec/input.json [hunk 39]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (osanseviero) [inline packages/tasks/src/tasks/zero-shot-object-detection/spec/input.json:1]: The output file is empty

File: packages/tasks/src/tasks/text-to-audio/inference.ts (hunk 33 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/text-to-audio/inference.ts:96]: I have added a bunch in 826181a - there are still a lot of other parameters to add

File: packages/tasks/src/tasks/token-classification/inference.ts (hunk 36 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/token-classification/inference.ts:29]: Ah you're right - but shouldn't be supported by the __call__ method too?

File: packages/tasks/src/tasks/video-classification/inference.ts (hunk 37 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/video-classification/inference.ts:44]: Yes - let's move the classification output to `schema-utils.json`

File: packages/tasks/src/tasks/text-to-audio/inference.ts (hunk 31 above)
Comment (SBrandeis) [reply packages/tasks/src/tasks/text-to-audio/inference.ts:29]: Let's do it as a follow-up

File: packages/tasks/src/tasks/token-classification/inference.ts (hunk 36 above)
Comment (osanseviero) [reply packages/tasks/src/tasks/token-classification/inference.ts:29]: I would expect yes, but maybe it changes how the architecture is loaded?

**Comment by Wauplin** [comment]
whoop whoop :rocket:
